An issue with async/timeout

There is raised an issue when the concurrency is high, and the concurrency distribution left much enough users to wait 200 milisecond to check again is it should run now. As the implementation of async/timeout uses timeouts map there is a case when one timeout channel instance is used for async take multiple times. Then an error is raised:
xception in thread "async-dispatch-216" java.lang.AssertionError: Assert failed: No more than 1024 pending takes are allowed on a single channel.
(< (.size takes) impl/MAX-QUEUE-SIZE)
	at clojure.core.async.impl.channels.ManyToManyChannel.take_BANG_(channels.clj:235)
	at clojure.core.async.impl.ioc_macros$take_BANG_.invokeStatic(ioc_macros.clj:983)
	at clojure.core.async.impl.ioc_macros$take_BANG_.invoke(ioc_macros.clj:982)
	at clj_gatling.simulation$run_scenario_constantly$fn__818$state_machine__5041__auto____847$fn__849.invoke(simulation.clj:116)
	at clj_gatling.simulation$run_scenario_constantly$fn__818$state_machine__5041__auto____847.invoke(simulation.clj:109)
	at clojure.core.async.impl.ioc_macros$run_state_machine.invokeStatic(ioc_macros.clj:973)
	at clojure.core.async.impl.ioc_macros$run_state_machine.invoke(ioc_macros.clj:972)
	at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invokeStatic(ioc_macros.clj:977)
	at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invoke(ioc_macros.clj:975)
	at clojure.core.async.impl.ioc_macros$take_BANG_$fn__5059.invoke(ioc_macros.clj:986)
	at clojure.core.async.impl.channels.ManyToManyChannel$fn__735.invoke(channels.clj:265)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

My test scenario is executed with concurrency 10k and concurrency-distribution 0.1. When I remove the distribution the problem with the timeout disappeared because no one user has to wait.

It is something close to my implementation.

(ns tests.timeout-issue
  (:require [clj-gatling.core :as gat]
            [taoensso.timbre :as timbre]
            [clojure.core.async :as a]
            [mount.core :as mount]))

(defn- setup
  "Initializing test context"
  [context]
  (timbre/info "Calling setup ...")
  (mount/start)
  (timbre/info "context: " context)
  context)

(defn- step-1 [context]
  (a/go
    (a/<! (a/timeout 200))))

(def step-2 step-1)
(def step-3 step-1)
(def step-4 step-1)
(def step-5 step-1)
(def step-6 step-1)
(def step-7 step-1)
(def step-8 step-1)

(defn- cleanup
  "Cleanup of test related data."
  [context]
  (timbre/info "Calling cleanup ...")
  (mount/stop)
  (timbre/info "Context cleaned up."))

(defn run-load [concurrency]
  (let [context {}]
    (gat/run {:name      "Simulation"
              :pre-hook  setup
              :post-hook cleanup
              :scenarios [{:name                     "Scenario"
                           :skip-next-after-failure? true   ;Optional (default true)
                           :allow-early-termination? true   ;Optional (default false)
                           :steps                    [{:name    "Step One"
                                                       :request step-1}
                                                      {:name    "Step Two"
                                                       :request step-2}
                                                      {:name    "Step Three"
                                                       :request step-3}
                                                      {:name    "Step Four"
                                                       :request step-4}
                                                      {:name    "Step Five"
                                                       :request step-5}
                                                      {:name    "Step Six"
                                                       :request step-6}
                                                      {:name    "Step Seven"
                                                       :request step-7}
                                                      {:name    "Step Eight"
                                                       :request step-8}
                                                      ]}]}
             {:concurrency              concurrency
              :concurrency-distribution (constantly 0.01)
              :timeout-in-ms            10000
              :context                  context}))
  )

(comment (run-load 10000))

And the error is the same as in my implementation:
Exception in thread "async-dispatch-2218" java.lang.AssertionError: Assert failed: No more than 1024 pending takes are allowed on a single channel.
(< (.size takes) impl/MAX-QUEUE-SIZE)
	at clojure.core.async.impl.channels.ManyToManyChannel.take_BANG_(channels.clj:235)
	at clojure.core.async.impl.ioc_macros$take_BANG_.invokeStatic(ioc_macros.clj:983)
	at clojure.core.async.impl.ioc_macros$take_BANG_.invoke(ioc_macros.clj:982)
	at clj_gatling.simulation$run_scenario_constantly$fn__818$state_machine__5041__auto____847$fn__849.invoke(simulation.clj:116)
	at clj_gatling.simulation$run_scenario_constantly$fn__818$state_machine__5041__auto____847.invoke(simulation.clj:109)
	at clojure.core.async.impl.ioc_macros$run_state_machine.invokeStatic(ioc_macros.clj:973)
	at clojure.core.async.impl.ioc_macros$run_state_machine.invoke(ioc_macros.clj:972)
	at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invokeStatic(ioc_macros.clj:977)
	at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invoke(ioc_macros.clj:975)
	at clojure.core.async.impl.ioc_macros$take_BANG_$fn__5059.invoke(ioc_macros.clj:986)
	at clojure.core.async.impl.channels.ManyToManyChannel$fn__735.invoke(channels.clj:265)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

--- a/src/clj_gatling/simulation.clj
+++ b/src/clj_gatling/simulation.clj
@@ -6,6 +6,7 @@
                                                  log-exception]]
             [schema.core :refer [validate]]
             [clj-time.local :as local-time]
+            [clj-gatling.timers :as timers]
             [clojure.core.async :as async :refer [go go-loop close! alts! <! >!]]))
 
 (set! *warn-on-reflection* true)
@@ -30,7 +31,7 @@
   (swap! sent-requests inc)
   (go
     (when-let [sleep-before (:sleep-before step)]
-      (<! (async/timeout (sleep-before original-context))))
+      (<! (timers/timeout (sleep-before original-context))))
     (let [original-context-with-user (assoc original-context :user-id user-id)
           start (now)
           return {:name (:name step)
@@ -38,7 +39,7 @@
                   :start start
                   :context-before original-context-with-user}
           response (asynchronize (:request step) original-context-with-user)
-          [{:keys [result end-time context exception]} c] (alts! [response (async/timeout timeout)])]
+          [{:keys [result end-time context exception]} c] (alts! [response (timers/timeout timeout)])]
       (if (= c response)
         [(assoc return :end end-time
                        :exception exception
@@ -134,7 +135,7 @@
                  (let [result (<! (run-scenario-once options scenario user-id))]
                    (swap! concurrent-scenarios dec)
                    (>! c result)))
-               (<! (async/timeout 200)))
+               (<! (timers/timeout 200)))
              (if (runners/continue-run? runner @sent-requests simulation-start)
                (recur)
                (close! c)))

--- a/src/clj_gatling/timers.clj
+++ b/src/clj_gatling/timers.clj
@@ -1,0 +1,55 @@
+(ns clj-gatling.timers
+  (:require [clojure.core.async.impl.protocols :as impl]
+            [clojure.core.async.impl.channels :as channels])
+  (:import [java.util.concurrent DelayQueue Delayed TimeUnit]))
+
+;Core.async has a limit of 1024 concurrent limits for waiting a same timeout channel
+;Internally core.async has a cache of timeout channels of same value
+;See: https://github.com/clojure/core.async/blob/master/src/main/clojure/clojure/core/async/impl/timers.clj
+;This is a version of timers.clj where all that caching has removed
+(set! *warn-on-reflection* true)
+
+(defonce ^:private ^DelayQueue timeouts-queue
+  (DelayQueue.))
+
+(deftype TimeoutQueueEntry [channel ^long timestamp]
+  Delayed
+  (getDelay [this time-unit]
+    (.convert time-unit
+              (- timestamp (System/currentTimeMillis))
+              TimeUnit/MILLISECONDS))
+  (compareTo
+   [this other]
+   (let [ostamp (.timestamp ^TimeoutQueueEntry other)]
+     (if (< timestamp ostamp)
+       -1
+       (if (= timestamp ostamp)
+         0
+         1))))
+  impl/Channel
+  (close! [this]
+    (impl/close! channel)))
+
+(defn- timeout-worker
+  []
+  (let [q timeouts-queue]
+    (loop []
+      (let [^TimeoutQueueEntry tqe (.take q)]
+        (impl/close! tqe))
+      (recur))))
+
+(defonce timeout-daemon
+  (delay
+   (doto (Thread. ^Runnable timeout-worker "clj-gatling.timers/timeout-daemon")
+     (.setDaemon true)
+     (.start))))
+
+(defn timeout
+  "returns a channel that will close after msecs"
+  [^long msecs]
+  @timeout-daemon
+  (let [timeout (+ (System/currentTimeMillis) msecs)
+        timeout-channel (channels/chan nil)
+        timeout-entry (TimeoutQueueEntry. timeout-channel timeout)]
+    (.put timeouts-queue timeout-entry)
+    timeout-channel))

